src/common/Validation.ml

```diff
@@ -63,7 +58,6 @@ end
   let apply x ~f = 
```

Comment:
(this comment isn't in quite in the right place, but doc above refers to `NonEmpty` list still).

Seems like there are a lot of ways you could encode the results of type checking. One way would be that we split Semantic_check into a type inference tree manipulation and a separate type checking phase that returns a possibly-empty list of `Either` Warnings or Errors; another might be to keep the phases intertwined but still use a single possibly empty list to represent warnings and error accumulation.

Can you tell me a little bit about how you thought about this design space?

Reply:
Type checking is inherently sequential e.g. we can't determine if function application is well typed until we have type checked each argument. We have to check types in sequence, from the bottom up.

Unless we do something a little bit fancy with error recovery, the type checker would stop as soon as it encountered the first error since it wouldn't have the necessary information to type check subsequent statements. So given e.g. a `Expr.Untyped.t` type checking would give us back a `(Expr.Typed.t,TypeError.t) Result.t`.

(If we _did_ implement error recovery, then we would always get a `Expr.Typed.t` but possibly a list of `TypeError.t`s.)

The point is that type checking will always need a monadic interface (whether implemented as a monad or not!) since it is sequential. Also, type checking moves us from `Untyped.t` to `Typed.t`, i.e. it decorates a piece of IR.

In contrast, semantic checking _should_ only capture the things we cannot check within the type system and may (currently does) depend on type information. At the moment (without an effect system) the things we can't verify with the type checker include things like illegal use of `TargetPlusEquals` outside the model block. 

The key thing here is that these semantic errors do not depend on one another - it is always illegal to place `TargetPlusEquals` in illegal places irrespective of other semantic errors. Semantic checking does not change the type of the intermediate representation.

Because one semantic check does not depend on a preceding one it only requires an applicative interface, so we can check for all errors in parallel. The semantic check should have the type `Expr.Typed.t -> (SemanticWarning.t list , (SemanticError.t * SemanticWarning.t) list) Result.t`.

If we were to introduce a simple effect system I think the only semantic errors we would check for would be 'identifier errors' and a couple of other things relating to duplicate names some of  which we could possibly do in the parser(?). 

Without an effect system, the additional errors would be around effects:

```ocaml
    | CannotAssignToReadOnly of string
    | CannotAssignToGlobal of string
    | ...
    | TargetPlusEqualsOutsideModelOrLogProb
```

(The first two reflect a function with an `Assign` effect and the last a `Target` effect. By checking different program blocks against sets of effects we can catch them as type errors.)

Either way I think separating type checking would be a good idea. Type systems can be hard enough to follow without mixing in unrelated logic.

Reply:
Thank you very much for writing this up! I had been under the impression that we could separate out a non-checking inference stage that returns a new tree, even if there were type errors. This could be a stretch goal one day as it might better support language servers as we could code inference such that it would essentially just use default logic to construct the tree - e.g. if two types don't line up, we leave those in the metadata incorrectly but then the parent expression appears well-typed from above. So inference is still monadic but perhaps checking becomes applicative and we can collect multiple type errors? Though maybe separating it out at that level would duplicate a ton of code. 

So I was wondering still about using basically `(Warning.t, Error.t) Either.t list` as the monad, rather than a custom one. Any thoughts about that?

Reply:
Sorry, need to think about the last question more... don't bother answering yet!

Reply:
Hmm okay I think I understand more and my design question translates to something like, why isn't `'a Validate.t = 'a * (Warning.t, Error.t) Either.t list)`? I see why, given that semantic checking is very entangled with inference now, that you need to always return a thing as well as the errors.  

Like not to get too Haskell-monad-transformers on you but we make a module that holds both some abstract type as well as some arbitrary monad and sort of does the obvious thing, piping the module's map to the underlying monad's map, etc. Does that make sense? 

Anyway I'd be happy to merge this once those three low level thingies are addressed, around pulling out some repeated functionality into a function and simplifying a bit